src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_eckey.c

```diff
@@ -0,0 +1,193 @@
+    {
+        oidStr = JSTRING("secp384r1");
+    }
+    else if (strcmp(oid, "1.3.132.0.35") == 0 || strcmp(oid, "nistP521") == 0)
+    {
+        oidStr = JSTRING("secp521r1");
+    }
+    else if (strcmp(oid, "1.2.840.10045.3.1.7") == 0 || strcmp(oid, "nistP256") == 0)
+    {
+        oidStr = JSTRING("secp256r1");
+    }
+    else 
+    {
+        oidStr = JSTRING(oid);
+    }
+    jstring ec = JSTRING("EC");
+
+    // First, generate the key pair based on the curve defined by the oid.
+    jobject paramSpec = (*env)->NewObject(env, g_ECGenParameterSpecClass, g_ECGenParameterSpecCtor, oidStr);
+    ReleaseLRef(env, oidStr);
+
+    jobject keyPairGenerator =
+        (*env)->CallStaticObjectMethod(env, g_keyPairGenClass, g_keyPairGenGetInstanceMethod, ec);
+    (*env)->CallVoidMethod(env, keyPairGenerator, g_keyPairGenInitializeWithParamsMethod, paramSpec);
+
+    ReleaseLRef(env, paramSpec);
+    if (CheckJNIExceptions(env))
+    {
+        LOG_DEBUG("Failed to create curve");
+        ReleaseLRef(env, ec);
+        ReleaseLRef(env, keyPairGenerator);
+        return NULL;
+    }
+
+    jobject keyPair = (*env)->CallObjectMethod(env, keyPairGenerator, g_keyPairGenGenKeyPairMethod);
+
+    ReleaseLRef(env, keyPairGenerator);
+
+    // Now that we have the key pair, we can get the curve parameters from the public key.
+    jobject keyFactory = (*env)->CallStaticObjectMethod(env, g_KeyFactoryClass, g_KeyFactoryGetInstanceMethod, ec);
+    jobject publicKey = (*env)->CallObjectMethod(env, keyPair, g_keyPairGetPublicMethod);
+    jobject keySpec =
+        (*env)->CallObjectMethod(env, keyFactory, g_KeyFactoryGetKeySpecMethod, publicKey, g_ECPublicKeySpecClass);
+
+    ReleaseLRef(env, ec);
+    ReleaseLRef(env, publicKey);
+    ReleaseLRef(env, keyFactory);
+
+    if (CheckJNIExceptions(env))
+    {
+        ReleaseLRef(env, keySpec);
+        ReleaseLRef(env, keyPair);
+        return NULL;
+    }
+
+    jobject curveParameters = (*env)->CallObjectMethod(env, keySpec, g_ECPublicKeySpecGetParams);
+
+    jobject group = (*env)->CallObjectMethod(env, curveParameters, g_ECParameterSpecGetCurve);
+
+    jobject field = (*env)->CallObjectMethod(env, group, g_EllipticCurveGetField);
+
+    ReleaseLRef(env, field);
+    ReleaseLRef(env, group);
+
+    return AndroidCryptoNative_NewEcKey(ToGRef(env, curveParameters), ToGRef(env, keyPair));
+}
+
+int32_t AndroidCryptoNative_EcKeyUpRef(EC_KEY* r)
+{
+    if (!r)
+        return FAIL;
+    r->refCount++;
+    return SUCCESS;
+}
+
+int32_t AndroidCryptoNative_EcKeyGetSize(const EC_KEY* key, int32_t* keySize)
+{
+    if (!keySize)
+        return FAIL;
+
+    *keySize = 0;
+
+    if (!key)
+        return FAIL;
+
+    JNIEnv* env = GetJNIEnv();
+
+    jobject curve = (*env)->CallObjectMethod(env, key->curveParameters, g_ECParameterSpecGetCurve);
+    jobject field = (*env)->CallObjectMethod(env, curve, g_EllipticCurveGetField);
+    *keySize = (*env)->CallIntMethod(env, field, g_ECFieldGetFieldSize);
```

Comment:
I don't remember how often we ask the native code for the keysize (vs trust that we loaded it correctly into the KeySize property).  If it's a lot, maybe this is something that we'd want to calculate once then pass in to NewEcKey (or calculate once inside NewEcKey).  If it's only once or so, the compartmentalization here is probably better.

Reply:
I think we most often just use the KeySize property.